crates/solver/src/settlement/external_prices.rs

```diff
@@ -0,0 +1,193 @@
+//! Module containing an external prices type use for converting amounts to
+//! their native asset value.
+//!
+//! Internally, the `ExternalPrices` keeps a set of exchange rates from tokens
+//! to the native asset and assumes the invariant that the exchange rate of the
+//! native asset and native wrapped token exist with a value of 1.
+
+use anyhow::{bail, Result};
+use ethcontract::{H160, U256};
+use lazy_static::lazy_static;
+use model::order::BUY_ETH_ADDRESS;
+use num::{BigInt, BigRational, One as _, ToPrimitive as _};
+use shared::conversions::U256Ext as _;
+use std::collections::{BTreeMap, HashMap};
+
+/// A collection of external prices used for converting token amounts to native
+/// assets.
+#[derive(Clone, Debug)]
+pub struct ExternalPrices(HashMap<H160, BigRational>);
+
+impl ExternalPrices {
+    /// Creates a new set of external prices for the specified exchange rates.
+    pub fn new(native_token: H160, mut xrates: HashMap<H160, BigRational>) -> Result<Self> {
+        // Make sure to verify our invariant that native asset price and native
+        // wrapped asset price exist with a value of 1. This protects us from
+        // malformed input (in case there are issues with the prices from the
+        // `/auction` endpoint for example).
+        for token in [native_token, BUY_ETH_ADDRESS] {
+            match xrates.get(&token) {
+                Some(price) if !price.is_one() => {
+                    let price = price.to_f64().unwrap_or(f64::NAN);
+                    bail!("malformed native token {:?} price {:.2e}", token, price);
+                }
+                Some(_) => {}
+                None => {
+                    xrates.insert(token, BigRational::one());
+                }
+            }
+        }
+
+        Ok(Self(xrates))
+    }
+
+    /// Returns a set of external prices for the specified auction model prices.
+    pub fn from_auction_prices(native_token: H160, prices: BTreeMap<H160, U256>) -> Result<Self> {
```

Comment:
nit: I think `try_from_action_prices` would be more fitting considering this returns a `Result`.